### ISSUE TITLE
Help wont open in Browser on Ubuntu System

### ISSUE THREAD

**Issue by BeFrankly** [helge17/tuxguitar#153]
Hello,

first I would say that it is great, that someone took the code and goes on with developing and fixing Tuxguitar. 

Yesterday I installed version 1.6.0. via dpkg on my Ubuntu 22.04.3 LTS and needed some help/documentation. So from the upper menu of tuxguitar i opened the help and so the standard browser, Firefox in my case, opened. The URL was file:///opt/tuxguitar/share/help/index.html and browser tells "error: file not found.".

In a terminal i did ls -l  /opt/tuxguitar/share/help and all the files are there and readable by owner, group and everyone. So why does it not open. Same with Chromium browser.

**Comment by guiv42** [comment]
This could be a limitation of Firefox, which version are you using?
See [there](https://askubuntu.com/questions/1435757/firefox-cant-open-files)

**Comment by BeFrankly** [comment]
Oh dear. I really need to get to grips with the configuration of Snap. Ok, security is an important point. 
I'll give it a try and report back if it works. Might take a while because I need to practise a few pieces on the guitar first. :-)

**Comment by BeFrankly** [comment]
Dear @guiv42 , I have tried the solution with apparmor from the link you provided. It does not work. But let's think again:
It is not a good idea to break Snap's security wall by editing the apparmor configuration just to see a documentation. In my opinion, it's better to provide a different path and maybe a different format of tuxguitar documentation that works with snap or other security restrictions.

**Comment by guiv42** [comment]
Hi @BeFrankly 
In fact I did not suggest to change this configuration, objective of this link was just to highlight that other Ubuntu users had similar issues. I just wanted to know if this could be the cause of your problem (I don't know either Ubuntu or Snap). Sorry if it was not clear. From your feedback I understand the problem is effectively there.
Anyway, the complete Ubuntu community will face this issue, so I agree it would be useful to find a solution.

Writing and maintaining doc is really time-consuming, and @helge17 already completely re-wrote this doc for 1.6.0. That would be for sure a significant amount of work to re-write it again. And static html format has the advantage to be easily deployed cross-platform (well... at least theoretically!). So I must admit I'm not enthusiastic about an evolution of format. Any idea about another possible format?

Another option could be to make this html content available on a web site (possible also to add a menu item "online help" targeting this site). 
But I would find this really frustrating. As you said "OK, security is an important point", so... forcing user to connect to the Internet to fetch some content already available locally... I'm not convinced it really increases security!

**Comment by BeFrankly** [comment]
Dear @guiv42 and @helge17 ,
thank you again for ongoing maintaining tuxguitar. To have the documentation in a structured format, i.e. html, is a good starting point. I guess that there a many apoaches about how solve the technical aspect of documentation. So html was your choise, but as I said, it's a good starting point. 
First, hosting the html doc on a website does cost nothing. Its just another entry in the help menu. Second, if you have html, what stops you converting it into pdf or ebook or what ever. Take a look at the software Calibre, a software for managing ebook libraries, and its capabilities in converting books.
There are also a lot of free libraries for converting docs. But you still have to decide, where to put the pdf-file on a ubuntu system. I don't know if the PDF reader is a snap app or not. But you can give it a try.

Regards

Frank

**Comment by guiv42** [comment]
> So html was your choise, but as I said, it's a good starting point.

To be fully honest, html was the historical solution used by TuxGuitar for years, so the original developer's choice (and it proved a quite efficient solution!)

> hosting the html doc on a website does cost nothing. Its just another entry in the help menu

That's not completely true. It implies some additional activities: administration/maintenance of the website, version management (so that user reaches the doc corresponding to the version of TuxGuitar he/she is using), etc. Basically, any duplication of data creates some additional work, and we are working with few resources.

Regarding other possible formats, I fear we might face exactly the same problem as soon as we rely on some external software to display the content: this software should have access to the local filesystem, and this cannot be guaranteed (as you said, a pdf reader could face the same problem than firefox, either today or tomorrow).

There is possibly another more ambitious option: to render directly html content within one TuxGuitar's window. I currently have a prototype working with SWT, so this could be a way forward. There is a similar feature available in JavaFX but I could not make it work yet. I'm currently investigating...

**Comment by guiv42** [comment]
Rendering html pages within TuxGuitar is clearly possible in both SWT and JFX configurations:

https://github.com/helge17/tuxguitar/assets/129443524/2211cafc-10f9-4467-b617-c8012448fef1

Advantages of this solution:
- fully independent from any external web browser, TuxGuitar displays help autonomously
- no network connection required for user
- no administration/maintenance of any website for developers/maintainers
- since it's available in both SWT and JFX this looks a standard feature, and it should probably be maintained

Drawbacks:
- some complexity added in TuxGuitar software (seems manageable)
- embedded browser is really basic, far from a full-featured firefox or chrome
- JFX only: new dependencies to JFX modules (pom files modified): javafx-web, javafx-media
- I dit not even consider Qt version (see #20)

So I can see 2 options to solve this issue in Ubuntu context:
1. Direct html rendering in TuxGuitar
2. Just add a menu item "online help" and provide help content online

@BeFrankly , @helge17 : what is your opinion?

And another question in case option 1 is preferred: would a (very basic) toolbar be required in the embedded browser? I think "back" and "forward" buttons might be useful, and would possibly be sufficient.

**Comment by BeFrankly** [comment]
Dear @guiv42 , it looks nice. So I don't care if it is just a simple browser plug-in with no features like Bookmarking and so on. It will be good, if one can search for terms within it. But if not, so what. As long as the the topics in the text are linked it would be easy to navigate through it.

We all know, that the best would be a context sensitive help-system, that opens the topic that belongs to the aspect that I'm in the app at that moment would be great. But look, it's a open source software no one pais for. It's up to you, how you solve it at the end.
By the way. Did you had contact to Julian Casadesus. Seems that his last commit at sourceforge was in 2022. So how comes that you fork the code to github? There a lot of links in the www pointing to an old URL, that's now working anymore. The URL is randomly redirected to some online shops. Horrible.

**Comment by guiv42** [comment]
OK, thanks for your feedback. As a first step, I will probably submit a pull request shortly, at least to restore existing functionality. That could be a good idea to automatically select an adapted web page depending from current context (to keep in mind for a future improvement)

> By the way. Did you had contact to Julian Casadesus. Seems that his last commit at sourceforge was in 2022. So how comes that you fork the code to github? There a lot of links in the www pointing to an old URL, that's now working anymore. The URL is randomly redirected to some online shops. Horrible.

You can find some history [here](https://sourceforge.net/p/tuxguitar/discussion/522984/thread/02743a8637/).
This fork was started by @helge17 (thanks again!), after trying to contact Julian without success.

**Comment by helge17** [comment]
The embedded HTML viewer looks very good, so I would vote for this solution if it's not too much work. I could also add some navigation buttons to the help pages in javascript if the embedded browser supports it.

But I can also host the help pages on tuxguitar.app. To do this, I need to extend my build script so that it automatically copies the latest documentation to the web server, including versioning etc. Not a big deal, but it still needs to be done.

I'm not sure what the best solution is in the long run. Everything is online these days. Musescore, for example, also links to its website from the help menu. I'm not sure how to deal with help pages on Android. Can they be displayed locally there? And of course someone has to take care of the online resources. I can do that in the foreseeable future, but obviously not forever. On the other hand, it's strange that you need an online connection just to see the help pages.

**Comment by guiv42** [comment]
Great, I like the idea of javascript for navigation controls. It's possible to do it above in TuxGuitar window, but it would imply to develop it twice (APIs differ for embedded browers in JFX and SWT).
So I will submit a pull request without the embedded controls, then we can try the 2 implementations in parallel.

About the online version, why not. We just have to take care to organize data on the website so several versions of the doc can co-exist, and link to the correct version in TuxGuitars' "online help". Long-term maintenance is definitely a challenge (see how the sites in .com.ar disappeared). So in any case the embedded doc shall be maintained I think.

**Comment by guiv42** [comment]
I just tested .deb packages (swt & jfx) produced by e103e01e0514da2a0d5c204459727d5b34215532 in a fresh install of Ubuntu 22.04, it works fine out of the box: doc is displayed in the embedded browser.
Just one limitation: I don't know why the newly created window does not include "maximize" control (it does on OpenSuse/KDE).
- with swt, double-click on the title banner of window maximizes it
- but not with javafx

Anyway, I would say this is acceptable (at least, doc is available)

**Comment by BeFrankly** [comment]
@guiv42 , do you have a wiki page that tells us how to build it ourselves? Is there any documentation for developers who want to help. What toolchain to use, what packages are needed. How can I test the DEB of the above mentioned commit.

**Comment by guiv42** [comment]
@helge17 did a great job to simplify and document the build process :)
First step is probably to build swt version of  TuxGuitar for your environment, detailed instructions are here: [install.md](https://github.com/helge17/tuxguitar/blob/master/TuxGuitar/doc/INSTALL.md).
Then, if you want to build the other variants (win/mac/... and JFX), you can take some inspiration from this [build script](https://github.com/helge17/tuxguitar/blob/master/misc/build_tuxguitar_from_source.sh).

To work on the code, I don't think anything is documented, and I could not find a perfect solution. I use "Eclipse IDE for java developers" (just import an existing maven project and browse to the pom file of linux/swt variant), but I did not succeed to build TuxGuitar with the native modules within Eclipse.

If you want to test without necessarily building TuxGuitar, pre-releases snapshot are published [here](https://github.com/helge17/tuxguitar/releases) quite often, the next one will include the embedded browser.

**Comment by BeFrankly** [comment]
So here I am back again. Yes, you did a good job. I'm so happy to have the full documentation working. This evening I first installed the deb for Jfx from the last night build. I could open and navigate through the doc. Fine :-)
The I installed the swt version, also working fine. At a first glance I did not recognize any differences between the two versions. Should there be any?

At a last step to finish this point I still think that placing the documentation online is still a good idea, because you can fix documentation bugs faster there than always have to roll out a new software version just because of a documentation bug.

Thanks a lot.

**Comment by guiv42** [comment]
> At a first glance I did not recognize any differences between the two versions. Should there be any?

Theoretically, no.
Julian has initiated this implementation with several ui frameworks (SWT + JFX + Qt), I did not find the rationale for this choice. Since this fork started here, we could not succeed to build the Qt version, but we are still maintaining both SWT + JFX. Until when, I don't know. Possibly one day we might choose to maintain only one.
Refer to [issues](https://github.com/helge17/tuxguitar/issues), there are some annoying bugs/limitations specific to JFX configuration, and one minor bug in SWT

**Comment by BeFrankly** [comment]
@guiv42 , sorry for my late reply. In the list of open issues I saw this is still open and was wondering why. Because of masses of email I have to mange, i did not recognize the mail about your entry. 
The issue can be closed now. You did a great job :+1: . 

But.... I will open a new issue today about another subject. It will be interesting, what you'll find out about that.

**Comment by drzraf** [comment]
> You can find some history [here](https://sourceforge.net/p/tuxguitar/discussion/522984/thread/02743a8637/). This fork was started by @helge17 (thanks again!), after trying to contact Julian without success.

Isn't this his LinkedIn profile ??
https://ar.linkedin.com/in/julian-casadesus
